keymaster/ta/attestation.c

```diff
@@ -162,7 +163,231 @@ TEE_Result TA_open_root_ec_attest_cert(TEE_ObjectHandle *attCert)
+	uint32_t			attrs_count;
+	uint64_t key_rsa_public_exponent = UNDEFINED;
+	uint32_t key_size = UNDEFINED;
+
+	DMSG("RSA root attestation key ...");
+
+	if (TA_decode_pkcs8(sessionSTA, key_data, &attrs,
+			&attrs_count, KM_ALGORITHM_RSA, &key_size,
+			&key_rsa_public_exponent) != KM_ERROR_OK)
+	{
+		goto error_1;
+	}
+
+	if (key_size % 8 != 0 || key_size > MAX_KEY_RSA) {
+		EMSG("RSA key size %d must be multiple of 8 and less than %u",
+							key_size,MAX_KEY_RSA);
+		goto error_2;
+	}
+
+	//Create object in storage
+	/* TODO : store in RPMB */
+	result = TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE,
+			RsaAttKeyID, sizeof(RsaAttKeyID),
+			TEE_DATA_FLAG_ACCESS_WRITE,
+			TEE_HANDLE_NULL, NULL, 0U, &RSAobject);
+	if (result != TEE_SUCCESS) {
+		EMSG("Failed to create a RSA persistent key, res=%x", result);
+		goto error_2;
+	}
+
+	for (uint32_t i = 0; i < attrs_count; i++) {
+		//Store RSA key in format: size | buffer attribute
+		DMSG("%s %d attrs[i].attributeID 0x%08X", __func__, __LINE__,attrs[i].attributeID);
+		result = TA_write_attest_obj_attr(RSAobject, attrs[i].content.ref.buffer, attrs[i].content.ref.length);
+		if (result != TEE_SUCCESS) {
+			EMSG("Failed to write RSA attribute %x, res=%x",
+					attrs[i].attributeID, result);
+			goto error_3;
+		}
+	}
+error_3:
+	(result == TEE_SUCCESS) ?
+			TEE_CloseObject(RSAobject) :
+			TEE_CloseAndDeletePersistentObject(RSAobject);
+error_2:
+	free_attrs(attrs, attrs_count);
+
+error_1:
+	return result;
+}
+
+static TEE_Result TA_set_ec_attest_key(TEE_TASessionHandle sessionSTA, keymaster_blob_t key_data)
+{
+	TEE_Result result = TEE_ERROR_BAD_PARAMETERS;
+	TEE_ObjectHandle ECobject = TEE_HANDLE_NULL;
+	TEE_Attribute *attrs;
+	uint32_t attrs_count;
+	uint64_t key_rsa_public_exponent = UNDEFINED;
+	uint32_t key_size = UNDEFINED;
+	uint32_t curve = UNDEFINED;
+
+	DMSG("EC root attestation key creation...");
+
+	if (TA_decode_pkcs8(sessionSTA, key_data, &attrs,
+			&attrs_count, KM_ALGORITHM_EC, &key_size,
+			&key_rsa_public_exponent) != KM_ERROR_OK)
+	{
+		goto error_1;
+	}
+
+	curve = TA_get_curve_nist(key_size);
+	if (curve == UNDEFINED) {
+		EMSG("Failed to get ECC curve nist");
+		result = TEE_ERROR_BAD_PARAMETERS;
+		goto error_2;
+	}
+
+	/* TODO : store in RPMB */
+	//Create object in storage
+	result = TEE_CreatePersistentObject(TEE_STORAGE_PRIVATE,
+			EcAttKeyID, sizeof(EcAttKeyID),
+			TEE_DATA_FLAG_ACCESS_WRITE,
+			TEE_HANDLE_NULL, NULL, 0U, &ECobject);
+	if (result != TEE_SUCCESS) {
+		EMSG("Failed to create a EC persistent key, res=%x", result);
+		goto error_2;
+	}
+
+	DMSG("%s %d curve 0x%08X", __func__, __LINE__,curve);
+	result = TEE_WriteObjectData(ECobject,
+			(void *)&curve, sizeof(uint32_t));
+	if (result != TEE_SUCCESS)
+	{
+		EMSG("Failed to write Curve value res=%x",
+				result);
+		goto error_3;
+	}
+
+	for (uint32_t i = 0; i < attrs_count; i++) {
+		//Attributes are "Ref"
+		DMSG("%s %d attrs[i].attributeID 0x%08X", __func__, __LINE__,attrs[i].attributeID);
```

Comment:
FYI, `DMSG()` already prints `__func__` and `__LINE__`, check `optee_os/lib/libutils/ext/include/trace.h`